### ISSUE TITLE
Unexpected mypy error for type-hints with generic types and inheritance

### ISSUE THREAD

**Issue by markuspinter** [python/mypy#15810]
**Bug Report**

Mypy complains when using Classes with Generics in dictionaries. More specifically I would like to use the type of the abstract class in the dict. See following example

**To Reproduce**

Gist URL: https://gist.github.com/mypy-play/2369116e3c026bb9ca39906ec4e2d5eb
Playground URL: https://mypy-play.net/?mypy=latest&python=3.10&gist=2369116e3c026bb9ca39906ec4e2d5eb

**Expected Behavior and Actual Behavior**

I would expect the first entry to be valid as well, since the Generic should be covariant and allow this subclass `Steakhouse(RestaurantBase[MeatDish])`, but it fails. 

```
main.py:69: error: The type "type[Steakhouse]" is not generic and not indexable  [misc]
Found 1 error in 1 file (checked 1 source file)
```

The only way I got it working is by either casting as in the second entry or by using Any in the type hint for the restaurants mapping like

```python
restaurants: Mapping[int, RestaurantBase[Any]]
```

**Your Environment**

- Mypy version used: 1.4.1
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10

In case you need more info, please let me know!

Thanks for your effort!

**Comment by hauntsaninja** [comment]
Your type variable is invariant, you likely want to make it covariant (although this does cause problems with `bar` in your example, not sure whether what you're doing is type safe)
https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types
